### ISSUE TITLE
Error: kine: building kine: failed to migrate DB from sqlite: COUNT(*)

### ISSUE THREAD

**Issue by jackielii** [ubuntu/microk8s#1604]
Cluster was running without any problem for weeks, then over the weekend it seems to have updated automatically? then it's starting to fail with:

```
...
microk8s.daemon-apiserver[1330333]: Error: kine: building kine: failed to migrate DB from sqlite: COUNT(*)
...
```

Seems the dqlite is totally gone? I'm running a 4 node HA cluster following the HA instructions: https://microk8s.io/docs/high-availability

[inspection-report-20200928_085251.tar.gz](https://github.com/ubuntu/microk8s/files/5291169/inspection-report-20200928_085251.tar.gz)

**Comment by ktsakalozos** [comment]
@freeekanayaka have you seen this before?

**Comment by freeekanayaka** [comment]
We would need a tarball of the `backend` directory to inspect. Note that I'm off this week so I'll be able to look at it only next week.

**Comment by ktsakalozos** [comment]
@jackielii the directory @freeekanayaka is asking for is `/var/snap/microk8s/current/var/kubernetes/backend/`. Would it be possible to create a tarball f it and attach it here? Node this is the kubernetes datastore.

**Comment by jackielii** [comment]
```
jackieli@r710-01:/var/snap/microk8s/current/var/kubernetes$ tar czvf ~/backend.tgz backend
backend/
backend/0000000075903276-0000000075903557
backend/0000000075900853-0000000075901127
backend/0000000075900429-0000000075900710
backend/0000000075895619-0000000075896109
backend/0000000075899838-0000000075899868
backend/0000000075897989-0000000075898254
backend/0000000075894615-0000000075895120
backend/0000000075901412-0000000075901682
backend/info.yaml
backend/cluster.key
backend/0000000075899536-0000000075899837
tar: backend/kine.sock: socket ignored
backend/0000000075896597-0000000075897014
backend/0000000075899869-0000000075900148
backend/0000000075903595-0000000075903599
backend/metadata2
backend/0000000075896110-0000000075896596
backend/0000000075903558-0000000075903585
backend/snapshot-141-75901994-2854845674.meta
backend/0000000075901683-0000000075901995
backend/snapshot-141-75901994-2854845674
tar: backend/snapshot-141-75901994-2854845674: file changed as we read it
backend/0000000075903600-0000000075903602
backend/0000000075898559-0000000075899040
backend/0000000075903586-0000000075903590
backend/0000000075897688-0000000075897988
backend/cluster.crt
backend/0000000075898255-0000000075898558
backend/0000000075899041-0000000075899535
backend/0000000075897015-0000000075897291
backend/0000000075900711-0000000075900852
backend/0000000075900149-0000000075900428
backend/0000000075902477-0000000075902968
backend/0000000075901996-0000000075902476
backend/0000000075903591-0000000075903594
backend/snapshot-142-75903018-2854864700
backend/0000000075895121-0000000075895618
backend/snapshot-142-75903018-2854864700.meta
backend/cluster.yaml
backend/0000000075897292-0000000075897687
backend/0000000075902969-0000000075903275
backend/metadata1
backend/0000000075901128-0000000075901411
tar: backend: file changed as we read it
```

file is 65MB, had to upload to gDrive: https://drive.google.com/file/d/1gZGNGxJjGJRUolfh91-TF0var7gJuwtQ/view?usp=sharing

**Comment by jackielii** [comment]
@ktsakalozos I thought I switched to stable channel after HA is available in 1.19/stable channel? so I followed the instructions and did: `sudo snap refresh microk8s --classic --channel=1.19/stable`

snap list:

```
...
microk8s  v1.19.2    1710   1.19/stable      canonical✓  classic
...
```

Is there a way for me to work around this problem?

**Comment by ktsakalozos** [comment]
I know very little of your setup. Are you willing to reinstall MicroK8s from `1.18/stable` until we fix this issue on the 1.19 channel. A clean re-install from `1.19` or `latest` channels may also address the issue.

**Comment by jackielii** [comment]
@ktsakalozos Thanks for replying. I have re-installed quite a few times and this time I thought it would be stable and installed my my dev work on top of it. it was about 1 weeks work to get it all up and running, so I'd rather fix the cluster instead of reinstall it all.

So if I understand correctly: 1.19/stable was updated with a Dqlite change and was automatically pushed to my cluster, and it broke my setup? I've seen people complaining about snaps. Now I feel it too

Not being negative etc. Let's fix the problem. I can provide all the info needed and wait for next week @freeekanayaka comes back to take a look. The deployments inside the cluster is still working, it's only I can't apply any updates to it. E.g. `k get no` returns:

```
❯ k get no
Error from server: rpc error: code = Unknown desc = (
                SELECT rkv.id
                FROM kine rkv
                ORDER BY rkv.id
                DESC LIMIT 1)
```

**Comment by ktsakalozos** [comment]
@jackielii I understand your frustration and I need to thank you for your patience.

With the 1.19 release we introduced dqlite as the kubernetes datastore. The issue you are seeing existed in the initial release and not introduced by an update, although I cannot rule out that a restart due to an update may have triggered it. The fact that we have automatic updates allows us to fix deployments when a patch is available.

You said you are running a 4 node cluster. All nodes are crashing? There is this doc [1] that explains how to recover a crashing node from a working one.

[1] https://discuss.kubernetes.io/t/recovery-of-ha-microk8s-clusters/12931

**Comment by jackielii** [comment]
Thanks for the link. Unfortunately, all nodes have the same issue. So I'm unable to recover the dqlite from a good node.

They're all on the same version. I tried `microk8s.status` and they all return `microk8s is not running. Use microk8s inspect for a deeper inspection.`

**Comment by freeekanayaka** [comment]
I've downloaded the tarball and tried to reproduce the problem locally, however the node starts fine here (I've changed the configuration to make it believe it's a single node, but the database will still be the same).

Do you still have this cluster and is it still broken in the same way? Did you try to restart all nodes yet?

**Comment by jackielii** [comment]
@freeekanayaka I have rebuilt the cluster using 1.18 already.

What I noticed is in the cluster.yaml:
```
- Address: 10.10.10.110:19001
  ID: 3297041220608546238
  Role: 2
- Address: 10.10.10.102:19001
  ID: 16993903095653174947
  Role: 0
- Address: 10.10.10.112:19001
  ID: 10648296310624360767
  Role: 0
- Address: 10.10.10.114:19001
  ID: 10808375087296616236
  Role: 1
- Address: 10.10.10.120:19001
  ID: 4080834572353121827
  Role: 0
```

"10.10.10.110" is not a active node it was an IP that's left the cluster. It's actually an old IP for 10.10.10.112. Although the strange thing is the IP change quite early in the process. It was running fine for quite a while, probably around 2 weeks.

**Comment by ktsakalozos** [comment]
Hi @jackielii,

how did you setup the 1.18 cluster? Did you snap refresh to the 1.18/stable channel or did you do a fresh installation with `snap install microk8s --channel=1.18/stable`? Could you attach a `microk8s inspect` tarball? I am asking because dqlite is not enabled by default on 1.18 therefore the cluster.yaml should not be used. I suspect you did a `snap refresh`.

The `cluster.yaml` keeps the list of nodes dqlite knows about. It seems the node at `10.10.10.110` was not disconnected from the cluster via `microk8s.leave` or `microk8s.remove-node`. Do you see this node as "NotReady" in `microk8s.kubectl get no` ?

**Comment by jackielii** [comment]
@ktsakalozos I think you misunderstood. I was describing the 1.19 installation. I fully teared down the 1.19 cluster and reinstalled 1.18. Now 1.18 is running fine.

I tried refresh from 1.19 to 1.18 first, but it seem to get stuck at some bad state. I remember running the status --wait command and it just timed out. I had to remove --purge and the install.

> The cluster.yaml keeps the list of nodes dqlite knows about. It seems the node at 10.10.10.110 was not disconnected from the cluster via microk8s.leave or microk8s.remove-node. Do you see this node as "NotReady" in microk8s.kubectl get no ?

I don't fully recall the exactly sequence, but I'm pretty sure the `kubect get no` didn't show the old IP. Could it be the new IP & old IP had the same hostname?
